### ISSUE TITLE
Broken battery controllers (handling module resets)

### ISSUE THREAD

**Issue by pstch** [polybar/polybar#1937]
## Is your feature request related to a problem? Please describe.

On my machine, the hardware exposes `charge_now` on the first boot, and after waking up from suspend, only `energy_now` is exposed. This makes `polybar` start watching `charge_now`, and it then fails to attach its inotify watch when this file has disappeared.

## Why does polybar need this feature?

This problem happens in the Thinkpad x220 and some other related models, and is very difficult to workaround without patching either the kernel or polybar. Patching polybar is accessible to me, working around this problem in the kernel is harder. The x220 is still a relatively popular laptop, and it would be nice if polybar could be used on it out-of-the-box.

## Describe the solution you'd like

When the `inotify` watch fails to reattach, we could "reset" the battery module and make it scan for sysfs files again. This wouldn't need any config option, except maybe to limit the amount of resets.

## Describe alternatives you've considered

 - patching the kernel and workaround the bug there (harder to patch, and if the source problem is in the battery module and not in its driver, it might not even be accepted).
 - entirely disabling the battery module, and using a script to show the battery status (not very practical)
 - restarting polybar each time I wake the laptop from suspend (this is the current solution I'm using)

## Additional context

 - I would totally understand if it's deemed not useful for polybar to handle such battery controllers. Restarting polybar is not a too bad workaround. This is why I submitted this as a feature request and not as a bug.
 - However, this "kind" of issue could happen in any module, so :
 - I'm submitting this issue so that it can be considered whether to have some "reset" logic for some modules. This would make sense for modules that have an "initialization" sequence, so that this initialization sequence can be triggered again.

**Comment by pstch** [comment]
Related bug reports:
 - https://github.com/polybar/polybar/issues/1181
 - https://github.com/polybar/polybar/issues/1046
 - https://github.com/polybar/polybar/issues/839 (only for users that get "0%")

Notes:
 - just allowing to choose between `charge_now` and `energy_now` would not be enough, as the machine seems to **switch** between one and the other. As polybar is already choosing between the two, making it *choose again* seems the easiest solution.
 - if the feature is accepted, I'll try to write the patch (I wanted to get some feedback before).

**Comment by patrick96** [comment]
Hi, thank you very much for offering to help out.

Indeed the battery module could use some work. We have a greater plan to specifically define the finite state machines for modules to make it easier to reason their behavior (see [here](https://github.com/polybar/polybar/projects/7)). There isn't anything concrete yet though. With that concept the battery module could have an error state that would then try again to reset itself. This would also involve something along the lines of #1529 with separating out the data gathering logic into an adapter.

We would very much appreciate it if you worked on this. If you want you can try to implement our state machine idea but I know that this would be more work and it would be fine if you "only" adapt the current module to reset itself.

**Comment by pstch** [comment]
Oh, it's nice to see that there are plans for this. Unfortunately, I'm not an experienced C/C++ developer, and wanted to implement the reset logic as a learning experience, so I don't think I would be able to handle implementing a state machine and data gathering adapters, at least not soon. I'm first going to go for the battery module reset logic, which may outline ideas and requirements for the FSM.

**Comment by patrick96** [comment]
Yes do that. Awesome that you are using this as a learning opportunity!

Let us know if you need anything

**Comment by f0s3** [comment]
I have the same problem on my x220. I may try to look into this, but as I am working a lot (as a software engineer, ofc), I'm not sure about when can I do this. If someone has looked into it already though, it may save some time to share some hands-on experience here (code snippets, descriptions of ideas for a proper fix/workaround).

**Comment by f0s3** [comment]
@patrick96, correct me if I'm wrong, but isn't this issue a bug? I'm about you putting a feature label on it.

**Comment by patrick96** [comment]
@f0s3 I think I mainly put the feature label on here because this was opened as a feature request. But it depends on how you look at it. It's all about how you expect the module to behave, if you think it should handle when its source of information no longer exists, then you would consider this a bug. I have never expected this from the module, but I also assumed that this doesn't really happen under normal circumstances.

In any case, I appreciate all the help I can get. I am also quite busy with polybar and IRL.

**Comment by patrick96** [comment]
Seems someone has already done this in #1306. I don't really have the time to review this right now, but maybe some of you can take a look at it.

**Comment by pstch** [comment]
Closing as this is fixed by #1306.

**Comment by Lomadriel** [comment]
We will close this issue when #1306 is merged.

**Comment by mafsi** [comment]
I have this problem too.
